keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,410 @@
+* Phase 2: Support of exact revision match.
+
+See also https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/365-paginated-lists#potential-future-extensions
+
+This change increases the K8s dependency on cache, the impact of a bug in caching logic is severe.
+Any incorrect behavior will be persisted in a local apiserver, and is extreamly hard to debug.
+As the proposed changes will in the end cover all apiserver list calls behind a cache, we need a better way to detect. 
+We propose to add a automatic mechanism for validating cache consistency with etcd,
+allowing users to trust it's results withount need to manually debug.
+
+### Snapshotting
+
+1. **Snapshot Creation:** When a watch event is received, the cacher creates
+   a snapshot of the B-tree based cache using the efficient [Clone()[] method.
+   This method creates a lazy copy of the tree structure, minimizing overhead.
+   Since the watch cache already stores the history of watch events,
+   the B-tree maintains just pointers to the in-use memory, storing only minimal necessary data.
+2. **Snapshot Storage:** Snapshots are stored in a separate tree data structure,
+   keyed by resourceVersion. This tree structure facilitates efficient lookup of
+   the "nextSmaller" element, as resourceVersions are not necessarily sequential.
+3. **Serving:** When a request requiring response based on previous snapshot arrives,
+   the API server performs the following steps:
+  - Extract the resourceVersion from request.
+  - Looks up the "nextSmaller" snapshot based on the resourceVersion.
+  - Constructs the response using data from the retrieved snapshot.
+  **Edge cases:**
+    - Requested resourceVersion is smaller than any available snapshot:
+      This indicates that the requested data has been cleaned up. In this scenario,
+      the API server falls back to serving the request from etcd.
+    - Requested resourceVersion is larger than the latest snapshot:
+      This could indicate a future resourceVersion or a situation where the watch
+      cache is lagging behind. The API server performs a consistent read from
+      etcd to confirm the existence of the future resourceVersion or waits for
+      the watch cache to catch up.
+4. **Snapshot Cleanup:** Snapshots are subject to a Time-To-Live (TTL) mechanism
+   similar to watch events. The proposed approach leverages the existing process
+   that limits the number of events to 10,000 within a 75-second window
+   (configurable via request timeout). Additionally, snapshots are purged during
+   cache re-initialization.
+
+[Clone()]: https://pkg.go.dev/github.com/google/btree#BTree.Clone
+
+### Cache Inconsistency detection
+
+For the first iteration of the mechanism we will calculate the hash based on LIST response
+at some specific RV of each resource and expose it as a metric. Hash will be calculated for
+both cache and etcd, allowing users to setup an alert to detect if they don't match.
```

Comment:
If this ever fails, I'd like the kube-apiserver to automatically revert to sending these requests to etcd and not handling them in the watch cache.  That leaves us a marker of "something broke" and avoids serving incorrect data.  I'm open a second gate (alpha and always off) that allows us to skip this behavior for diagnostic purposes, but we should not knowing continue to use a broken feature in production clusters.

Reply:
I'm worried about introducing an untested automatic switch mechanism that would cause severe performance impact. For Alpha I was planning to make the inconsistency detection default (assuming overhead is acceptable) to qualify it.

As for what action we should take if "something broke" I was thinking more about purging cache. It's more risky as it would make resource unavailable, that's why we should first qualify the inconsistency detection.

Reply:
+1 to Marek
We definitely need concistency-checker and I would like it to be enabled in all our CI pretty much from the beginning.
As for what to do in prod - given how significant the different between serving from cache vs etcd is, automatic fallback to etcd may have very bad consequences. Purging the cache and recreating it from scratch sounds much less risky to me.

Reply:
If the mechanism is producing an inconsistent cache, I don't agree that simply restarting that inconsistent cache is a good way to resolve the problem.

Reply:
Discussed offline, we are in agreement that fallback to etcd would be better here, at least until GA.